curator-framework/src/main/java/org/apache/curator/framework/state/CircuitBreaker.java

```diff
@@ -0,0 +1,79 @@
+package org.apache.curator.framework.state;
+
+import org.apache.curator.RetryPolicy;
+import org.apache.curator.RetrySleeper;
+import java.util.Objects;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+// must be guarded by sync
+class CircuitBreaker
+{
+    private final RetryPolicy retryPolicy;
+    private final ScheduledExecutorService service;
+
+    private boolean isOpen = false;
+    private int retryCount = 0;
+    private long startNanos = 0;
+
+    CircuitBreaker(RetryPolicy retryPolicy, ScheduledExecutorService service)
+    {
+        this.retryPolicy = Objects.requireNonNull(retryPolicy, "retryPolicy cannot be null");
+        this.service = Objects.requireNonNull(service, "service cannot be null");
+    }
+
+    boolean isOpen()
+    {
+        return isOpen;
+    }
+
+    int getRetryCount()
+    {
+        return retryCount;
+    }
+
+    boolean tryToOpen(Runnable completion)
+    {
+        if ( isOpen )
+        {
+            return false;
+        }
+
+        isOpen = true;
+        retryCount = 0;
+        startNanos = System.nanoTime();
+        if ( tryToRetry(completion) )
+        {
+            return true;
+        }
+        close();
+        return false;
+    }
+
+    boolean tryToRetry(Runnable completion)
+    {
+        if ( !isOpen )
+        {
+            return false;
+        }
+
+        long[] sleepTimeNanos = new long[]{0L};
+        RetrySleeper retrySleeper = (time, unit) -> sleepTimeNanos[0] = unit.toNanos(time);
+        if ( retryPolicy.allowRetry(retryCount, System.nanoTime() - startNanos, retrySleeper) )
```

Comment:
Note: this is a little clumsy given how Curator's RetryPolicy works. But, this is getting the amount of sleep the retry desires (internally the retry policy calls the sleeper with a sleep amount)

Reply:
According to the documentation of `RetryPolicy.allowRetry`, the second argument is supposed to be in milliseconds. You provide nanos instead.

Reply:
Java's `TimeUnit` can convert to any unit. So, for future benefit I'm converting to nanos via `unit.toNanos`.

Reply:
If I read the code correctly, it can cause problems with policies like `RetryUntilElapsed`. Please correct me if I am wrong

Reply:
Doh - I misread your comment. I see the problem now and it's fixex.